skorch/tests/test_probabilistic.py

```diff
@@ -0,0 +1,697 @@
+
+    The method/attribute required to determine the shape depends on the kind of
+    distrubtion.
+
+    """
+    shape = getattr(dist, 'shape', None)
+    if shape:
+        return shape[0]
+
+    get_base_samples = getattr(dist, 'get_base_samples', None)
+    if get_base_samples:
+        return get_base_samples().shape[0]
+
+    shape = getattr(dist, 'batch_shape', None)
+    if shape:
+        return shape[0]
+
+    raise AttributeError(f"Could not determine shape of {dist}")
+
+
+# PyTorch Modules are defined on the module root to make them pickleable.
+
+class RbfModule(gpytorch.models.ExactGP):
+    """Simple exact GP regression module"""
+    def __init__(self, X, y, likelihood):
+        super().__init__(X, y, likelihood)
+        self.mean_module = gpytorch.means.ConstantMean()
+        self.covar_module = gpytorch.kernels.RBFKernel()
+
+    def forward(self, x):
+        mean_x = self.mean_module(x)
+        covar_x = self.covar_module(x)
+        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
+
+
+class VariationalRegressionModule(gpytorch.models.ApproximateGP):
+    """GP regression for variational inference"""
+    def __init__(self, inducing_points, eps=1e-6):
+        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
+            inducing_points.size(0))
+        variational_strategy = gpytorch.variational.VariationalStrategy(
+            self,
+            inducing_points,
+            variational_distribution,
+            learn_inducing_locations=True,
+        )
+        super().__init__(variational_strategy)
+        self.mean_module = gpytorch.means.ConstantMean()
+        self.covar_module = gpytorch.kernels.ScaleKernel(
+            gpytorch.kernels.RBFKernel(eps=eps))
+
+    def forward(self, x):
+        mean_x = self.mean_module(x)
+        covar_x = self.covar_module(x)
+        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
+
+
+class VariationalBinaryClassificationModule(gpytorch.models.ApproximateGP):
+    """GP classification for variational inference"""
+    def __init__(self, inducing_points, eps=1e-6):
+        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
+            inducing_points.size(0))
+        variational_strategy = gpytorch.variational.UnwhitenedVariationalStrategy(
+            self,
+            inducing_points,
+            variational_distribution,
+            learn_inducing_locations=False,
+        )
+        super().__init__(variational_strategy)
+        self.mean_module = gpytorch.means.ConstantMean()
+        self.covar_module = gpytorch.kernels.ScaleKernel(
+            gpytorch.kernels.RBFKernel(eps=eps))
+
+    def forward(self, x):
+        mean_x = self.mean_module(x)
+        covar_x = self.covar_module(x)
+        latent_pred = gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
+        return latent_pred
+
+
+class BaseProbabilisticTests:
```

Comment:
I was doing an experiment here. Do you think this overall improves the code or not? It does reduce code duplication but it may be less obvious what is going on.

Reply:
It's mostly okay with me. Looking at this more closely, it all sense and its a reasonable way to "do not repeat yourself". I can see other approaches, but I do not see it as drastic improvement over what we have now.